vxquery-core/src/main/java/org/apache/vxquery/jsonparser/JSONParser.java

```diff
@@ -189,12 +385,139 @@ public void atomicValues(int tag, JsonParser parser, DataOutput out, StringValue
+                }
+            }
+
+        }
+        if (count == 2 && valueSeq != null) {
+            this.arrayCounters.add(Integer.valueOf(0));
+        }
+    }
+
+    public void startArray() throws HyracksDataException {
+        levelArray++;
+        if (levelArray > abStack.size()) {
+            abStack.add(new ArrayBuilder());
+        }
+        if (levelArray + levelObject > abvsStack.size() - 1) {
+            abvsStack.add(new ArrayBackedValueStorage());
+        }
+        startArrayOrObjects(2);
+        itemStack.add(itemType.ARRAY);
+        if (this.pathMatch() || this.valueSeq == null) {
+            abvsStack.get(levelArray + levelObject).reset();
+            try {
+                abStack.get(levelArray - 1).reset(abvsStack.get(levelArray + levelObject));
+            } catch (Exception e) {
+                throw new HyracksDataException("Accessing index out of bounds", e);
+            }
+        }
+    }
+
+    public void startObject() throws HyracksDataException {
+        levelObject++;
+        if (levelObject > obStack.size()) {
+            obStack.add(new ObjectBuilder());
+        }
+        if (levelArray + levelObject > abvsStack.size() - 1) {
+            abvsStack.add(new ArrayBackedValueStorage());
+        }
+        startArrayOrObjects(1);
+        itemStack.add(itemType.OBJECT);
+        if (this.pathMatch() || this.valueSeq == null) {
+            abvsStack.get(levelArray + levelObject).reset();
+            try {
+                obStack.get(levelObject - 1).reset(abvsStack.get(levelArray + levelObject));
+            } catch (Exception e) {
+                throw new HyracksDataException("Accessing index out of bounds", e);
             }
         }
     }
+
+    public void startFieldName(JsonParser parser) throws HyracksDataException {
+        if (levelObject > spStack.size()) {
+            keyStack.add(new ArrayBackedValueStorage());
+            spStack.add(new UTF8StringPointable());
+        }
+        keyStack.get(levelObject - 1).reset();
+        DataOutput outk = keyStack.get(levelObject - 1).getDataOutput();
+        try {
+            svb.write(parser.getText(), outk);
+            spStack.get(levelObject - 1).set(keyStack.get(levelObject - 1));
+            if (this.valueSeq != null) {
+                int length = 0;
+                byte[] barr = spStack.get(levelObject - 1).getByteArray();
+                outputStream.reset();
+                outputStream.write(barr, 0, spStack.get(levelObject - 1).getLength());
+                allKeys.add(ArrayUtils.toObject(outputStream.toByteArray()));
+                for (int i = 0; i < allKeys.size() - 1; i++) {
+                    tvp.set(ArrayUtils.toPrimitive(allKeys.get(i)), 0, ArrayUtils.toPrimitive(allKeys.get(i)).length);
+                    length += ArrayUtils.toPrimitive(allKeys.get(i)).length;
+                }
+                if (length <= pathStream.size() && (length + 2) <= pathStream.size()) {
```

Comment:
Why 2? How can we make this more clear what this value represents? Comment and/or turning this into a constant with a descriptive name.